File: src/util/main/sp/it/util/access/ref/LazyR.java
package sp.it.util.access.ref;

import java.util.function.Consumer;
import java.util.function.Supplier;
import org.jetbrains.annotations.Nullable;
import sp.it.util.functional.Functors.F1;
import static sp.it.util.dev.FailKt.noNull;

/**
 * Nullable lazy value.
 *
 * @param <V> type of instance
 */
public class LazyR<V> {

	protected V v;
	protected Supplier<V> builder;
	protected boolean isSet = false;

	/**
	 * @param builder produces the instance when it is first accessed. Builder can produce null. It will never execute
	 * more than once.
	 */
	public LazyR(Supplier<V> builder) {
		this.builder = noNull(builder);
	}

	public V get() {
		if (!isSet) set(builder.get());
		return v;
	}

	protected void set(V val) {
		isSet = true;
		builder = null;
		v = val;
	}

	public <M> V get(M m, F1<M,V> or) {
		if (!isSet) set(or.apply(m));
		return v;
	}

	/** @return the current value if set, otherwise the specified value */
	public V getOr(V or) {
		return isSet ? v : or;
	}

	/** @return the current value if set, otherwise null */
	public @Nullable V getOrNull() {
		return getOr((V) null);
	}

	/** @return the current value if set, otherwise the specified value */
	public V getOr(Supplier<V> or) {
		return isSet ? v : or.get();
	}

	public boolean isSet() {
		return isSet;
	}

	public void ifSet(Consumer<? super V> block) {
		if (isSet)
			block.accept(v);
	}

}

File: src/util/main/sp/it/util/access/ref/LazyR.kt
package sp.it.util.access.ref

import java.util.function.Consumer
import java.util.function.Supplier

/**
 * Nullable lazy value.
 * @param builder produces the instance when it is first accessed. Builder can produce null.
 * @param <V> type of instance
 */
@Suppress("UNCHECKED_CAST")
open class LazyR<V>(builder: () -> V) {
   protected var v: V? = null
   protected var builder: (() -> V)? = builder
   var isInitialized: Boolean = false
      protected set

   protected fun set(value: V) {
      isInitialized = true
      builder = null
      v = value
   }

   fun get(): V {
      if (!isInitialized) set(builder!!())
      return v as V
   }

   fun <M> get(m: M, or: (M) -> V): V {
      if (!isInitialized) set(or(m))
      return v as V
   }

   /** @return the current value if set, otherwise the specified value */
   fun getOr(or: V): V =
      if (isInitialized) v as V else or

   /** @return the current value if set, otherwise null */
   val orNull: V?
      get() = if (isInitialized) v else null

   /** @return the current value if set, otherwise the specified value */
   fun getOr(or: Supplier<V>): V =
      if (isInitialized) v as V else or.get()

   fun ifInitialized(block: Consumer<in V>) {
      if (isInitialized) block.accept(v as V)
   }

}